File: 131-palindrome-partitioning/131-palindrome-partitioning.java
class Solution {
    public List<List<String>> partition(String s) {
        List<List<String>>res=new ArrayList<>();
        solve(0,s,res,new ArrayList<>());
        return res;
    }
    public void solve(int ind,String s,List<List<String>>res,List<String>temp){
        int n=s.length();
        if(ind>=n){
            res.add(new ArrayList<>(temp));
        }
        for(int i=ind;i<s.length();i++){
            if(isok(ind,i,s)){
                String t=s.substring(ind,i+1);
                temp.add(t);
                solve(i+1,s,res,temp);
                temp.remove(temp.size()-1);
            }
        }
    }
    public boolean isok(int start,int end,String s){
        while(start<=end){
            if(s.charAt(start)==s.charAt(end)){
                start++;
                end--;
            }else
                return false;
        }
        return true;
    }
}

File: 131-palindrome-partitioning/131-palindrome-partitioning.cpp
class Solution {
public:
    vector<vector<string>>res;
    bool isok(string s,int ind,int i){
        while(ind<i){
            if(s[ind]!=s[i])
                return false;
            ind++;
            i--;
        }
        return true;
    }
    void solve(int ind,string s,vector<string>
& temp){
        if(ind==s.size()){
            res.push_back(temp);
            return ;
        }
        for(int i=ind;i<s.size();i++){
            if(isok(s,ind,i)){
                string p=s.substr(ind,i-ind+1);
                temp.push_back(p);
                solve(i+1,s,temp);
                temp.pop_back();
            }
        }
    }
    vector<vector<string>> partition(string s) {
        vector<string>temp;
        solve(0,s,temp);
        return res;
    }
};